From 850286ae52e4b2041d17b63488607779a36bfcab Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Thu, 2 Nov 2023 22:41:03 +0100
Subject: [PATCH 0060/1695] plat-stm32mp1: syscfg: HLSV mode for IO domains

Add platform API functions stm32mp_set_hslv_state() and
stm32mp_enable_fixed_vdd_hslv() to configure High Speed Low Voltage
mode of IO domains.

Platform function stm32mp_enable_fixed_vdd_hslv() is designed for
fixed voltage IO domains that need to be enable at boot time only
since the supply voltage level never changes.

On STM32MP13 variants, SDMMC IO domains may not be supplied by fixed
voltage VDD but rather by a supply which voltage level can change
at runtime for example to support SD/MMC normative 1.8V and 3.3V voltage
modes. Therefore these IO domains require a runtime configuration
function implemented by stm32mp_set_hslv_state().

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 4a65ce49956af93978e72d6b8e635e37f04a4814 plat-stm32mp1: syscfg: HLSV mode for IO domains]
Change-Id: I0c0d77f2d6d65638c28118164b6786a9b8235adb
Co-developed-by: Pascal Paillet <p.paillet@foss.st.com>
Signed-off-by: Pascal Paillet <p.paillet@foss.st.com>
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
[etienne: cherry-picked from 43e0957ad66eef409607e885865afdb78a9121e6]
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
---
 .../plat-stm32mp1/drivers/stm32mp1_syscfg.c   | 71 ++++++++++++++++++-
 .../plat-stm32mp1/drivers/stm32mp1_syscfg.h   | 27 +++++++
 2 files changed, 95 insertions(+), 3 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.c b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.c
index 8f795caf3..1b2ecbcff 100644
--- a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.c
+++ b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.c
@@ -18,12 +18,23 @@
 /*
  * SYSCFG register offsets (base relative)
  */
+#define SYSCFG_IOCTRLSETR			U(0x18)
 #define SYSCFG_CMPCR				U(0x20)
 #define SYSCFG_CMPENSETR			U(0x24)
 #define SYSCFG_CMPSD1CR				U(0x30)
 #define SYSCFG_CMPSD2CR				U(0x40)
+#define SYSCFG_HSLVEN0R				U(0x50)
 #define SYSCFG_IOSIZE				U(0x400)
 
+/*
+ * SYSCFG_IOCTRLSETR Register for STM32MP15 variants
+ */
+#define SYSCFG_IOCTRLSETR_HSLVEN_TRACE		BIT(0)
+#define SYSCFG_IOCTRLSETR_HSLVEN_QUADSPI	BIT(1)
+#define SYSCFG_IOCTRLSETR_HSLVEN_ETH		BIT(2)
+#define SYSCFG_IOCTRLSETR_HSLVEN_SDMMC		BIT(3)
+#define SYSCFG_IOCTRLSETR_HSLVEN_SPI		BIT(4)
+
 /*
  * SYSCFG_CMPCR Register
  */
@@ -44,6 +55,12 @@
  */
 #define SYSCFG_CMPENSETR_MPU_EN			BIT(0)
 
+/*
+ * HSLV definitions
+ */
+#define SYSCFG_HSLV_MASK			GENMASK_32(15, 0)
+#define SYSCFG_HSLV_KEY				U(0x1018)
+
 static vaddr_t get_syscfg_base(void)
 {
 	static struct io_pa_va base = { .pa = SYSCFG_BASE };
@@ -85,16 +102,16 @@ static __maybe_unused void disable_io_compensation(int cmpcr_offset)
 	      value_cmpcr2 & SYSCFG_CMPENSETR_MPU_EN))
 		return;
 
-	/* Copy apsrc / ansrc in ransrc /rapsrc */
+	/* Copy APSRC (resp. ANSRC) in RAPSRC (resp. RANSRC) */
 	apsrc_ansrc = value_cmpcr >> SYSCFG_CMPCR_ANSRC_SHIFT;
 	value_cmpcr &= ~(SYSCFG_CMPCR_RANSRC | SYSCFG_CMPCR_RAPSRC);
 	value_cmpcr |= SHIFT_U32(apsrc_ansrc, SYSCFG_CMPCR_RANSRC_SHIFT);
 
 	io_write32(cmpcr_base, value_cmpcr | SYSCFG_CMPCR_SW_CTRL);
 
-	DMSG("SYSCFG.cmpcr = %#"PRIx32, io_read32(cmpcr_base));
-
 	io_setbits32(cmpcr_base + CMPENCLRR_OFFSET, SYSCFG_CMPENSETR_MPU_EN);
+
+	DMSG("SYSCFG.cmpcr = %#"PRIx32, io_read32(cmpcr_base));
 }
 
 static TEE_Result stm32mp1_iocomp(void)
@@ -131,4 +148,52 @@ void stm32mp_set_vddsd_comp_state(enum stm32mp13_vddsd_comp_id id, bool enable)
 	else
 		disable_io_compensation(cmpcr_offset);
 }
+
+void stm32mp_set_hslv_state(enum stm32mp13_hslv_id id, bool enable)
+{
+	size_t hslvenxr_offset = 0;
+	uint32_t hlvs_value = 0;
+
+	assert(id < SYSCFG_HSLV_COUNT);
+
+	if (enable)
+		hlvs_value = SYSCFG_HSLV_KEY;
+
+	/* IDs are indices of SYSCFG_HSLVENxR registers */
+	hslvenxr_offset = SYSCFG_HSLVEN0R + id * sizeof(uint32_t);
+
+	io_write32(get_syscfg_base() + hslvenxr_offset, hlvs_value);
+
+	/* Value read shall be 1 on enable and 0 on disable */
+	hlvs_value = io_read32(get_syscfg_base() + hslvenxr_offset) &
+		     SYSCFG_HSLV_MASK;
+	if (enable != hlvs_value)
+		panic();
+}
+
+void stm32mp_enable_fixed_vdd_hslv(void)
+{
+	enum stm32mp13_hslv_id id = SYSCFG_HSLV_COUNT;
+
+	for (id = SYSCFG_HSLV_IDX_TPIU; id < SYSCFG_HSLV_COUNT; id++) {
+		/* SDMMCs domains may not be supplied by VDD */
+		if (id == SYSCFG_HSLV_IDX_SDMMC1 ||
+		    id == SYSCFG_HSLV_IDX_SDMMC2)
+			continue;
+
+		stm32mp_set_hslv_state(id, true);
+	}
+}
 #endif /* CFG_STM32MP13 */
+
+#ifdef CFG_STM32MP15
+void stm32mp_enable_fixed_vdd_hslv(void)
+{
+	io_write32(get_syscfg_base() + SYSCFG_IOCTRLSETR,
+		   SYSCFG_IOCTRLSETR_HSLVEN_TRACE |
+		   SYSCFG_IOCTRLSETR_HSLVEN_QUADSPI |
+		   SYSCFG_IOCTRLSETR_HSLVEN_ETH |
+		   SYSCFG_IOCTRLSETR_HSLVEN_SDMMC |
+		   SYSCFG_IOCTRLSETR_HSLVEN_SPI);
+}
+#endif
diff --git a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.h b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.h
index 2d35da2eb..575f147be 100644
--- a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.h
+++ b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.h
@@ -6,6 +6,23 @@
 #ifndef __DRIVERS_STM32MP1_SYSCFG_H
 #define __DRIVERS_STM32MP1_SYSCFG_H
 
+/* High Speed Low Voltage domains IDs for STM32MP13 variants */
+enum stm32mp13_hslv_id {
+	SYSCFG_HSLV_IDX_TPIU = 0,
+	SYSCFG_HSLV_IDX_QSPI,
+	SYSCFG_HSLV_IDX_ETH1,
+	SYSCFG_HSLV_IDX_ETH2,
+	SYSCFG_HSLV_IDX_SDMMC1,
+	SYSCFG_HSLV_IDX_SDMMC2,
+	SYSCFG_HSLV_IDX_SPI1,
+	SYSCFG_HSLV_IDX_SPI2,
+	SYSCFG_HSLV_IDX_SPI3,
+	SYSCFG_HSLV_IDX_SPI4,
+	SYSCFG_HSLV_IDX_SPI5,
+	SYSCFG_HSLV_IDX_LTDC,
+	SYSCFG_HSLV_COUNT
+};
+
 /* IO compensation domains IDs for STM32MP13 variants */
 enum stm32mp13_vddsd_comp_id {
 	SYSCFG_IO_COMP_IDX_SD1,
@@ -20,5 +37,15 @@ enum stm32mp13_vddsd_comp_id {
  * @enable: True to enable IO compensation, false to disable
  */
 void stm32mp_set_vddsd_comp_state(enum stm32mp13_vddsd_comp_id id, bool enable);
+
+/*
+ * Enable or disable High Speed Low Voltage mode of an IO domain
+ * @index: HSLV IO domain ID
+ * @enable: True to enable IO compensation, false to disable
+ */
+void stm32mp_set_hslv_state(enum stm32mp13_hslv_id id, bool enable);
 #endif /*CFG_STM32MP13*/
+
+/* Enable High Speed Low Voltage mode for domains fixed supplied VDD */
+void stm32mp_enable_fixed_vdd_hslv(void);
 #endif /*__DRIVERS_STM32MP1_SYSCFG_H*/
-- 
2.47.2

